specs/file_hierarchy_for_the_verification_of_os_artifacts.md

```diff
@@ -0,0 +1,561 @@
+
+Users of a VOA hierarchy (such as package installation software) can pick the relevant set of _signature verifiers_ for the verification of a specific artifact based on their location in the hierarchy.
+
+### Load paths
+
+VOA defines a list of _load paths_ for system mode and user mode, from which a hierarchy of _signature verifiers_ can be retrieved.
+
+Load paths when in system mode:
+
+- `/etc/voa/`
+- `/run/voa/`
+- `/usr/local/share/voa/`
+- `/usr/share/voa/`
+
+Load paths when in user mode:
+
+- `$XDG_RUNTIME_DIR/voa/`
+- `$XDG_DATA_HOME/voa/`
+
+Refer to the [XDG Base Directory Specification] for defaults.
+
+_Signature verifiers_ are retrieved from these directories to validate artifacts based on specific [load logic].
+This is referred to as _verifier lookup_.
+
+The existence of multiple load paths allows system administrators and users to define custom sets of _signature verifiers_ for their system and applications.
+
+### Symlinking
+
+Symlinks can be used in the VOA hierarchy to point to files or directories below one of the [load paths].
+Symlinking to files or directories external to the load paths is not supported.
+
+### Masking
+
+Individual _artifact verifiers_ and _trust anchors_ may be masked using a symlink to `/dev/null`, independent of [technology].
+This constitutes the only allowed exception to the rule of not symlinking to files or directories external to the load paths.
+
+These symlinks are only expected to be used in writable and persistent load paths (that is, only below `/etc/voa/` when in system mode, or below`$XDG_DATA_HOME/voa/` when in user mode).
+Masked verifiers are not considered when retrieving from any of the VOA [load paths].
+
+Masking _artifact verifiers_ and _trust anchors_ may be desirable in situations where a verifier is considered compromised, but a revocation for it can not be provided to users in time.
+VOA offers this mechanism as an alternative to revocation.
+
+### Load logic
+
+Load logic depends on the [technology] in use.
+It is the mechanism by which relevant verifier information from all load paths is retrieved, while considering the [os], [purpose], [context] and [technology] request of a calling application.
+
+For some technologies, the first suitable _signature verifier_ found in the list of [load paths] is used, effectively [overriding] files with the same name in directories lower in the list.
+
+However, in other technologies, more complex [merging] logic may apply, that effectively combines information from different representations of one verifier.
+
+Additionally, the concept of [masking] applies for all technologies.
+
+#### Overriding
+
+In technologies that use overriding, verifier files shadow other variants of the same verifier based on filename equality.
+
+For example, `/etc/voa/fedora/packages/default/ssh/buildserver1.pub` overrides `/usr/share/voa/fedora/package/default/ssh/buildserver1.pub`.
```

Comment:
You do not define which directories have what priority. Therefore, you can end up with various implementations.

Reply:
Do you think it should be made more clear than what is mentioned in https://github.com/uapi-group/specifications/pull/134/files#diff-c79d1da1ef2fcfffc28ac7308505535eac0942d086d54a990553374ac81fed00R148?

Granted, the priority is not mentioned explicitly in load paths or load logic, albeit that would probably be good indeed.

Reply:
nod. E.g. what has higher priority: `/usr/share/voa/` or `/etc/voa/`?

Reply:
migth make sense to reference https://uapi-group.org/specifications/specs/configuration_files_specification/

that spec suggests /etc overrides both /run and /usr, and /run overriding /usr

Reply:
Good point. Makes sense to weave that in!

Reply:
Added in https://github.com/uapi-group/specifications/pull/134/commits/20dcf7432b17d7bca04905eb3ddc8c64122e84fb